module/zfs/zfs_ioctl.c

```diff
@@ -3651,6 +3681,48 @@ zfs_ioc_bookmark(const char *poolname, nvlist_t *innvl, nvlist_t *outnvl)
 	return (dsl_bookmark_create(innvl, outnvl));
 }
 
+
+/*
+ * Clone existing bookmarks. Bookmark names are of the form <fs>#<bmark>.
+ * All bookmarks must be in the same pool.
+ *
+ * innvl: {
+ *     new_bookmark1 -> source_bm1, new_bookmark2 -> source_bm2
+ * }
+ *
+ * outnvl: bookmark -> error code (int32)
+ *
+ */
+static const zfs_ioc_key_t zfs_keys_bookmark_clone[] = {
+	{"<bookmark>...",	DATA_TYPE_STRING,	ZK_WILDCARDLIST},
+};
+
+/* ARGSUSED */
+static int
+zfs_ioc_bookmark_clone(const char *poolname, nvlist_t *innvl, nvlist_t *outnvl)
```

Comment:
What's the reasoning behind creating a new ioctl, vs. allowing additional types of arguments (source can be a bookmark) to the existing ioctl?

Regardless of the user-kernel interface, I think that we should be able to reuse more of the zfs_ioc_bookmark() code for the new functionality.  i.e. modify `dsl_bookmark_create_*` to handle the source being a bookmark.  You would probably make it convert the source snapshot to a bookmark_phys_t, as we do for zfs send.  See handling of `ancestor_zb` in `dmu_send()`:

```
		zfs_bookmark_phys_t *zb = &dspp.ancestor_zb;
		if (strchr(fromsnap, '@') != NULL) {
			err = dsl_dataset_hold(dspp.dp, fromsnap, FTAG,
			    &fromds);
				if (!dsl_dataset_is_before(dspp.to_ds, fromds,
				    0)) {
					err = SET_ERROR(EXDEV);
				} else {
					ASSERT3U(dspp.is_clone, ==,
					    (dspp.to_ds->ds_dir !=
					    fromds->ds_dir));
					zb->zbm_creation_txg =
					    dsl_dataset_phys(fromds)->
					    ds_creation_txg;
...
		} else {
			dspp.numfromredactsnaps = NUM_SNAPS_NOT_REDACTED;
			err = dsl_bookmark_lookup(dspp.dp, fromsnap, dspp.to_ds,
			    zb);

```

Reply:
Both the decision to create a new ioctl and the decision to have completely separate code paths for creation vs cloning were made because

* I wasn't clear about how much overlap there actually is between source_type=snapshot vs source_type=bookmark
* I did not know whether people would like the feature to live in the `zfs clone` vs `zfs bookmark` subcommand
* type-dependent branching deep down in the check tree makes it harder to reason about the correctness of the checks due to the increased amount of possible control flows

However, I see the point that a lot of the `dsl_bookmark_clone_*` stuff is boilerplate.
Thanks for the hint to the `zfs send` code.
I will merge the cloning functionality into the `dsl_bookmark_create_*` machinery.